pandas/core/frame.py

```diff
@@ -6482,7 +6482,32 @@ def combiner(x, y):
 
             return expressions.where(mask, y_values, x_values)
 
-        return self.combine(other, combiner, overwrite=False)
+        combined = self.combine(other, combiner, overwrite=False)
+
+        dtypes = {}
+
+        for col in self.columns.intersection(other.columns):
+            try:
+                # if the column has different dtype in the
+                # DataFrame objects then add the common dtype
+                # to the columns dtype conversion dict
+                if combined.dtypes[col] != self.dtypes[col]:
+                    dtypes[col] = find_common_type(
+                        [self.dtypes[col], other.dtypes[col]]
+                    )
+            except TypeError:
```

Comment:
we do not want to do multiple try/excepts ever as these tend to hide errors.
in fact you should *not* need here at all. `find_common_type` will always succeed (it could of course be `object`).